src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+                   .ToArray();
+            }
+            else
+            {
+                addresses = new[] {result};
+            }
+            
+            if (!addresses.Any())
+            {
+                var msg = "Cannot determine address(es) for " + result;
+                
+                foreach (var ip in ips)
+                {
+                    msg += " nic-ip: " + ip.Address;
+                }
+                
+                cancel.Cancel();
+                throw new Exception(msg);
+            }
+
+            // Add actual addresses to listeners and local peer addresses.
+            foreach (var a in addresses)
+            {
+                _log.Debug($"Listening on {a}");
+                _listeners.TryAdd(a, cancel);
+            }
+
+            LocalPeer.Addresses = LocalPeer
+               .Addresses
+               .Union(addresses)
+               .ToArray();
+
+            ListenerEstablished?.Invoke(this, LocalPeer);
+            return Task.FromResult(addresses.First());
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///   Called when a remote peer is connecting to the local peer.
+        /// </summary>
+        /// <param name="stream">
+        ///   The stream to the remote peer.
+        /// </param>
+        /// <param name="local">
+        ///   The local peer's address.
+        /// </param>
+        /// <param name="remote">
+        ///   The remote peer's address.
+        /// </param>
+        /// <remarks>
+        ///   Establishes the protocols of the connection.  Any exception is simply
+        ///   logged as warning.
+        /// </remarks>
+        private async void OnRemoteConnect(Stream stream, MultiAddress local, MultiAddress remote)
+#pragma warning restore VSTHRD100 // Avoid async void methods
+        {
+            if (!IsRunning)
+            {
+                try
+                {
+                    stream.Dispose();
+                }
+                catch (Exception)
+                {
+                    // eat it.
+                }
+
+                return;
+            }
+
+            // If the remote is already trying to establish a connection, then we
+            // can just refuse this one.
+            if (!_pendingRemoteConnections.TryAdd(remote, null))
+            {
+                _log.Debug($"Duplicate remote connection from {remote}");
+                try
+                {
+                    stream.Dispose();
+                }
+                catch (Exception)
+                {
+                    // eat it.
+                }
+
+                return;
+            }
+
+            try
+            {
+                if (_log.IsDebugEnabled)
+                {
+                    _log.Debug($"remote connect from {remote}");
+                }
+
+                // TODO: Check the policies
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)